src/main/java/org/corfudb/runtime/clients/LayoutClient.java

```diff
@@ -111,10 +111,10 @@ public void handleMessage(CorfuMsg msg, ChannelHandlerContext ctx) {
      *              Otherwise, the completablefuture completes exceptionally
      *              with OutrankedException.
      */
-    public CompletableFuture<LayoutPrepareResponse> prepare(long rank)
+    public CompletableFuture<LayoutPrepareResponse> prepare(Layout layout, long rank)
     {
         return router.sendMessageAndGetCompletable(
-                new LayoutRankMsg(null, rank, CorfuMsgType.LAYOUT_PREPARE)
+                new LayoutRankMsg(layout, rank, CorfuMsgType.LAYOUT_PREPARE)
```

Comment:
As @slfritchie mentions, you could insert into the message only an epoch instead of the full layout object